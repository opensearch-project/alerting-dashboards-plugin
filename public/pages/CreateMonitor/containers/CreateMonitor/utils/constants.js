/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  WHERE_FILTER_ALLOWED_TYPES,
  OPERATORS_MAP,
} from '../../../components/MonitorExpressions/expressions/utils/constants';
import { MONITOR_TYPE } from '../../../../../utils/constants';
import { SUPPORTED_DOC_LEVEL_QUERY_OPERATORS } from '../../../components/DocumentLevelMonitorQueries/utils/constants';

export const BUCKET_COUNT = 5;

export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);

export const FORMIK_INITIAL_WHERE_EXPRESSION_VALUES = {
  fieldName: [{ label: '', type: WHERE_FILTER_ALLOWED_TYPES[0] }], // This is an array because the EuiCombobox returns an array of {label: string, type: string} objects.
  operator: OPERATORS_MAP.IS.value,
  fieldValue: '',
  fieldRangeStart: undefined,
  fieldRangeEnd: undefined,
};

/** Sample delegate
 *  {
      order: 1,
      monitor_id: '{{m1}}',
    }
 */
export const DEFAULT_ASSOCIATED_MONITORS_VALUE = {
  sequence: {
    delegates: [],
  },
};

export const FORMIK_INITIAL_VALUES = {
  /* CONFIGURE MONITOR */
  name: '',
  disabled: false,
  frequency: 'interval',
  timezone: [],
  daily: 0,
  period: { interval: 1, unit: 'MINUTES' },
  weekly: { mon: false, tue: false, wed: false, thur: false, fri: false, sat: false, sun: false },
  monthly: { type: 'day', day: 1 },
  cronExpression: '0 */1 * * *',

  /* DEFINE MONITOR */
  monitor_type: MONITOR_TYPE.QUERY_LEVEL,
  searchType: 'graph',
  clusterNames: [],
  uri: {
    api_type: '',
    clusters: [],
    path: '',
    path_params: '',
    url: '',
  },
  index: [],
  timeField: '',
  query: MATCH_ALL_QUERY,
  queries: [],
  description: '',
  aggregationType: 'count',
  fieldName: [],
  aggregations: [],
  groupBy: [],
  groupByField: [{ label: '' }],
  overDocuments: 'all documents',
  groupedOverTop: 5,
  groupedOverFieldName: 'bytes',
  bucketValue: 1,
  bucketUnitOfTime: 'h', // m = minute, h = hour, d = day
  filters: [], // array of FORMIK_INITIAL_WHERE_EXPRESSION_VALUES
  detectorId: '',
  associatedMonitors: DEFAULT_ASSOCIATED_MONITORS_VALUE,
  associatedMonitorsList: [],
  associatedMonitorsEditor: '',
  preventVisualEditor: false,
};

export const FORMIK_INITIAL_AGG_VALUES = {
  aggregationType: 'count',
  fieldName: '',
};

export const FORMIK_INITIAL_DOCUMENT_LEVEL_QUERY_VALUES = {
  id: undefined,
  queryName: '',
  field: '',
  operator: SUPPORTED_DOC_LEVEL_QUERY_OPERATORS[0],
  query: '',
  tags: [],
};

// TODO DRAFT: Is this an appropriate default to display when defining as an extraction query?
export const DEFAULT_DOCUMENT_LEVEL_QUERY = JSON.stringify(
  {
    description: 'DESCRIPTION_TEXT',
    queries: [
      {
        name: 'QUERY_NAME',
        query: { match_all: {} },
        tags: ['TAG_TEXT'],
      },
    ],
  },
  null,
  4
);

export const DEFAULT_COMPOSITE_AGG_SIZE = 50;

export const METRIC_TOOLTIP_TEXT = 'Extracted statistics such as simple calculations of data.';
export const TIME_RANGE_TOOLTIP_TEXT = 'The time frame of data the plugin should monitor.';
export const FILTERS_TOOLTIP_TEXT =
  'Use a filter to retrieve a subset of the original data source.';
export const GROUP_BY_TOOLTIP_TEXT = 'Specify a field whose every unique value can trigger alerts.';
