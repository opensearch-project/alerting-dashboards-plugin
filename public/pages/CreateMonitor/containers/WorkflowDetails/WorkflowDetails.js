/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ContentPanel from '../../../../components/ContentPanel';
import Schedule from '../../components/Schedule';
import AssociateMonitors from '../../components/AssociateMonitors/AssociateMonitors';
import { EuiSpacer } from '@elastic/eui';

const WorkflowDetails = ({ values, isDarkMode, httpClient, errors }) => {
  return (
    <ContentPanel
      title="Workflow"
      titleSize="s"
      panelStyles={{
        paddingBottom: '20px',
        paddingLeft: '10px',
        paddingRight: '10px',
        paddingTop: '20px',
      }}
    >
      <Schedule isAd={false} />

      <EuiSpacer size="xl" />
      <AssociateMonitors
        isDarkMode={isDarkMode}
        values={values}
        httpClient={httpClient}
        errors={errors}
      />
    </ContentPanel>
  );
};

export default WorkflowDetails;
