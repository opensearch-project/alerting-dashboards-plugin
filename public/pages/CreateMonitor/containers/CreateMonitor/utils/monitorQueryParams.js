/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { SEARCH_TYPE } from '../../../../../utils/constants';

export const initializeFromQueryParams = (queryParams) => {
  return {
    searchType: queryParams.searchType || undefined,
    name: queryParams.name ? `${queryParams.name}-Monitor` : undefined,
    detectorId: queryParams.adId || undefined,
    period:
      queryParams.interval && queryParams.unit
        ? { interval: parseInt(queryParams.interval), unit: queryParams.unit }
        : undefined,
    adResultIndex: queryParams.adResultIndex || undefined,
  };
};
