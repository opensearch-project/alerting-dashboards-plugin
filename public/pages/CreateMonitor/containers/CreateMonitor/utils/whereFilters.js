/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';
import { DATA_TYPES } from '../../../../../utils/constants';

/**
 * Breakdown to factory pattern for rules in-case we support multiple filters.
 * This is just ease for the single one.
 */
export const OPERATORS_QUERY_MAP = {
  [OPERATORS_MAP.IS.value]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? { match_phrase: { [label]: fieldValue } }
        : { term: { [label]: fieldValue } },
  },
  [OPERATORS_MAP.IS_NOT.value]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            bool: { must_not: { match_phrase: { [label]: fieldValue } } },
          }
        : {
            bool: { must_not: { term: { [label]: fieldValue } } },
          },
  },
  [OPERATORS_MAP.IS_NULL.value]: {
    query: ({ fieldName: [{ label: fieldKey }] }) => ({
      bool: { must_not: { exists: { field: fieldKey } } },
    }),
  },
  [OPERATORS_MAP.IS_NOT_NULL.value]: {
    query: ({ fieldName: [{ label: fieldKey }] }) => ({ exists: { field: fieldKey } }),
  },
  [OPERATORS_MAP.IS_GREATER.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { gt: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { gte: fieldValue } },
    }),
  },
  [OPERATORS_MAP.IS_LESS.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { lt: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IS_LESS_EQUAL.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { lte: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IN_RANGE.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldRangeStart, fieldRangeEnd }) => ({
      range: { [fieldKey]: { gte: fieldRangeStart, lte: fieldRangeEnd } },
    }),
  },
  [OPERATORS_MAP.NOT_IN_RANGE.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldRangeStart, fieldRangeEnd }) => ({
      bool: { must_not: { range: { [fieldKey]: { gte: fieldRangeStart, lte: fieldRangeEnd } } } },
    }),
  },

  [OPERATORS_MAP.STARTS_WITH.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      prefix: { [fieldKey]: fieldValue },
    }),
  },

  [OPERATORS_MAP.ENDS_WITH.value]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      wildcard: { [fieldKey]: `*${fieldValue}` },
    }),
  },
  [OPERATORS_MAP.CONTAINS.value]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            query_string: { query: `*${fieldValue}*`, default_field: label },
          }
        : {
            wildcard: { [label]: `*${fieldValue}*` },
          },
  },
  [OPERATORS_MAP.DOES_NOT_CONTAINS.value]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            bool: {
              must_not: { query_string: { query: `*${fieldValue}*`, default_field: label } },
            },
          }
        : {
            bool: { must_not: { wildcard: { [label]: `*${fieldValue}*` } } },
          },
  },
};
