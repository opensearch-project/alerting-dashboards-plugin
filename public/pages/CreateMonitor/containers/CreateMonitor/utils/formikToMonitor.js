/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import moment from 'moment-timezone';
import { BUCKET_COUNT, DEFAULT_COMPOSITE_AGG_SIZE, FORMIK_INITIAL_VALUES } from './constants';
import { MONITOR_TYPE, SEARCH_TYPE } from '../../../../../utils/constants';
import { OPERATORS_QUERY_MAP } from './whereFilters';
import {
  API_TYPES,
  URL_DEFAULT_PREFIX,
} from '../../../components/ClusterMetricsMonitor/utils/clusterMetricsMonitorConstants';
import {
  getApiPath,
  getApiType,
} from '../../../components/ClusterMetricsMonitor/utils/clusterMetricsMonitorHelpers';
import {
  COMPOSITE_INPUT_FIELD,
  DOC_LEVEL_INPUT_FIELD,
  DOC_LEVEL_QUERY_MAP,
} from '../../../components/DocumentLevelMonitorQueries/utils/constants';

export function formikToMonitor(values) {
  const uiSchedule = formikToUiSchedule(values);
  const schedule = buildSchedule(values.frequency, uiSchedule);

  const monitorUiMetadata = () => {
    switch (values.monitor_type) {
      case MONITOR_TYPE.DOC_LEVEL:
        return {
          [DOC_LEVEL_INPUT_FIELD]: formikToDocLevelQueriesUiMetadata(values),
          search: { searchType: values.searchType },
        };
      case MONITOR_TYPE.COMPOSITE_LEVEL:
        return {
          [COMPOSITE_INPUT_FIELD]: formikToCompositeUiMetadata(values),
          search: { searchType: values.searchType },
        };
      default:
        return { search: formikToUiSearch(values) };
    }
  };

  if (values.monitor_type === MONITOR_TYPE.COMPOSITE_LEVEL) {
    const enabled_time = new Date();
    return {
      last_update_time: enabled_time.getTime(),
      owner: 'alerting',
      type: 'workflow',
      enabled_time: enabled_time.getTime(),
      enabled: !values.disabled,
      monitor_type: MONITOR_TYPE.COMPOSITE_LEVEL,
      workflow_type: MONITOR_TYPE.COMPOSITE_LEVEL,
      schema_version: 0,
      name: values.name,
      schedule,
      inputs: [formikToInputs(values)],
      triggers: [],
      ui_metadata: {
        schedule: uiSchedule,
        monitor_type: values.monitor_type,
        ...monitorUiMetadata(),
      },
    };
  }

  return {
    name: values.name,
    type: 'monitor',
    monitor_type: values.monitor_type,
    enabled: !values.disabled,
    schedule,
    inputs: [formikToInputs(values)],
    triggers: [],
    ui_metadata: {
      schedule: uiSchedule,
      monitor_type: values.monitor_type,
      ...monitorUiMetadata(),
    },
  };
}

export function formikToInputs(values) {
  switch (values.monitor_type) {
    case MONITOR_TYPE.CLUSTER_METRICS:
      return formikToClusterMetricsInput(values);
    case MONITOR_TYPE.DOC_LEVEL:
      return formikToDocLevelInput(values);
    case MONITOR_TYPE.COMPOSITE_LEVEL:
      return formikToCompositeInput(values);
    default:
      return formikToSearch(values);
  }
}

export function formikToCompositeInput(values) {
  return {
    composite_input: values.associatedMonitors,
  };
}

export function formikToSearch(values) {
  const isAD = values.searchType === SEARCH_TYPE.AD;
  let query = isAD ? formikToAdQuery(values) : formikToQuery(values);
  const adResultIndex = _.get(values, 'adResultIndex', '.opendistro-anomaly-results*');
  const indices = isAD ? [adResultIndex] : formikToIndices(values);

  return {
    search: {
      indices,
      query,
    },
  };
}

export function formikToAdQuery(values) {
  return {
    size: 1,
    sort: [{ anomaly_grade: 'desc' }, { confidence: 'desc' }],
    query: {
      bool: {
        filter: [
          {
            range: {
              execution_end_time: {
                from: '{{period_end}}||-' + values.period.interval + 'm',
                to: '{{period_end}}',
                include_lower: true,
                include_upper: true,
              },
            },
          },
          {
            term: {
              detector_id: {
                value: values.detectorId,
              },
            },
          },
        ],
      },
    },
    aggregations: {
      max_anomaly_grade: {
        max: {
          field: 'anomaly_grade',
        },
      },
    },
  };
}

export function formikToClusterMetricsInput(values) {
  let apiType = _.get(values, 'uri.api_type', FORMIK_INITIAL_VALUES.uri.api_type);
  if (_.isEmpty(apiType)) apiType = getApiType(_.get(values, 'uri'));
  let pathParams = _.get(values, 'uri.path_params', FORMIK_INITIAL_VALUES.uri.path_params);
  pathParams = _.trim(pathParams);
  // Trim '/' characters from the beginning and end of the path
  pathParams = pathParams?.replace(/^\/+|\/+$/g, '');
  const hasPathParams = !_.isEmpty(pathParams);
  const path = getApiPath(hasPathParams, apiType);
  let url = FORMIK_INITIAL_VALUES.uri.url;
  if (!_.isEmpty(apiType)) {
    url = URL_DEFAULT_PREFIX;
    if (!_.isEmpty(path)) url = url + '/' + path;
    if (hasPathParams) {
      pathParams = '/' + pathParams;
      url = url + pathParams + _.get(API_TYPES, `${apiType}.appendText`, '');
    }
  }
  const clusterNames = _.get(values, 'clusterNames', []);

  return {
    uri: {
      api_type: apiType,
      path: path,
      path_params: pathParams,
      url: url,
      clusters: clusterNames,
    },
  };
}

export function formikToAd(values) {
  return {
    anomaly_detector: {
      detector_id: values.detectorId,
    },
  };
}

export function formikToUiSearch(values) {
  const { searchType, timeField, aggregations, groupBy, bucketValue, bucketUnitOfTime, filters } =
    values;
  const cleanedGroupBy = groupBy.filter((item) => item !== '');
  return {
    searchType,
    timeField,
    aggregations,
    groupBy: cleanedGroupBy,
    bucketValue,
    bucketUnitOfTime,
    filters,
  };
}

export function formikToIndices(values) {
  const hasRemoteClusters = values.index.some(
    ({ cluster, value }) => !_.isEmpty(cluster) && !_.isEmpty(value)
  );
  return values.index.map(({ label, value }) => (hasRemoteClusters ? value : label));
}

export function formikToQuery(values) {
  const isGraph = values.searchType === SEARCH_TYPE.GRAPH;
  return isGraph ? formikToGraphQuery(values) : formikToExtractionQuery(values);
}

export function formikToExtractionQuery(values) {
  let query = _.get(values, 'query', FORMIK_INITIAL_VALUES.query);
  try {
    // JSON.parse() throws an exception when the argument is a malformed JSON string.
    // This caused exceptions when tinkering with the JSON in the code editor.
    // This try/catch block will only parse the JSON string if it is not malformed.
    // It will otherwise store the JSON as a string for continued editing.
    query = JSON.parse(query);
  } catch (err) {}
  return query;
}

export function formikToGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime, monitor_type } = values;

  const aggregation = () => {
    switch (monitor_type) {
      case MONITOR_TYPE.BUCKET_LEVEL:
        return formikToCompositeAggregation(values);
      default:
        return formikToAggregation(values);
    }
  };
  const timeField = values.timeField;
  let filters = [
    {
      range: {
        [timeField]: {
          gte: `{{period_end}}||-${Math.round(bucketValue)}${bucketUnitOfTime}`,
          lte: '{{period_end}}',
          format: 'epoch_millis',
        },
      },
    },
  ];
  const whereFilters = formikToWhereClause(values);
  if (whereFilters.length) filters = filters.concat(whereFilters);
  return {
    size: 0,
    aggregations: aggregation(),
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToDocLevelInput(values) {
  let description = FORMIK_INITIAL_VALUES.description;
  let indices = formikToIndices(values);
  let queries = _.get(values, 'queries', FORMIK_INITIAL_VALUES.queries);
  switch (values.searchType) {
    case SEARCH_TYPE.GRAPH:
      description = values.description;
      queries = queries.map((query) => {
        const formikToQuery = DOC_LEVEL_QUERY_MAP[query.operator].query(query);
        return {
          id: query.id,
          name: query.queryName,
          query: formikToQuery,
          tags: query.tags,
        };
      });
      break;
    case SEARCH_TYPE.QUERY:
      let query = _.get(values, 'query', '');
      try {
        query = JSON.parse(query);
        description = _.get(query, 'description', description);
        queries = _.get(query, 'queries', queries);
      } catch (e) {
        /* Ignore JSON parsing errors as users may just be configuring the query */
      }
      break;
    default:
      console.log(
        `Unsupported searchType found for ${MONITOR_TYPE.DOC_LEVEL}: ${JSON.stringify(
          values.searchType
        )}`,
        values.searchType
      );
  }

  return {
    [DOC_LEVEL_INPUT_FIELD]: {
      description: description,
      indices: indices,
      queries: queries,
    },
  };
}

export function formikToDocLevelQueriesUiMetadata(values) {
  return { queries: _.get(values, 'queries', []) };
}

export function formikToCompositeUiMetadata(values) {
  return {
    associatedMonitors: _.get(values, 'associatedMonitors', []),
    query: _.get(values, '', ''),
  };
}

export function formikToCompositeAggregation(values) {
  const { aggregations, groupBy } = values;

  let aggs = {};
  aggregations.map((aggItem) => {
    // TODO: Changing any occurrence of '.' in the fieldName to '_' since the
    //  bucketSelector uses the '.' syntax to resolve aggregation paths.
    //  Should revisit this as replacing with `_` could cause collisions with fields named like that.
    const name = `${aggItem.aggregationType}_${aggItem.fieldName.replace(/\./g, '_')}`;
    const type = aggItem.aggregationType === 'count' ? 'value_count' : aggItem.aggregationType;
    aggs[name] = {
      [type]: { field: aggItem.fieldName },
    };
  });
  let sources = [];
  groupBy.map((groupByItem) =>
    sources.push({
      [groupByItem]: {
        terms: {
          field: groupByItem,
        },
      },
    })
  );
  return {
    composite_agg: {
      composite: { sources },
      aggs,
    },
  };
}

// For aggregations of query-level monitor
export function formikToAggregation(values) {
  const { aggregations, groupBy } = values;

  let aggs = {};
  aggregations.map((aggItem) => {
    const type = aggItem.aggregationType === 'count' ? 'value_count' : aggItem.aggregationType;
    aggs['metric'] = {
      [type]: { field: aggItem.fieldName },
    };
  });
  if (groupBy.length)
    aggs.terms_agg = {
      terms: {
        field: groupBy[0],
      },
    };
  return aggs;
}

/////////// Build Graph UI Search Request ///////////////

export function formikToUiGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime, monitor_type } = values;
  const useComposite = monitor_type === MONITOR_TYPE.BUCKET_LEVEL;
  const aggregation = useComposite
    ? formikToUiCompositeAggregation(values)
    : formikToUiOverAggregation(values);
  const timeField = values.timeField;
  let filters = [
    {
      range: {
        [timeField]: {
          // default range window to [BUCKET_COUNT] * the date histogram interval
          gte: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          lte: 'now',
        },
      },
    },
  ];
  const whereFilters = formikToWhereClause(values);
  if (whereFilters.length) filters = filters.concat(whereFilters);
  return {
    size: 0,
    aggregations: aggregation,
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToUiOverAggregation(values) {
  const metricAggregation = formikToMetricAggregation(values);
  const { bucketValue, bucketUnitOfTime, groupBy } = values;
  const timeField = values.timeField;
  const aggregations = groupBy.length
    ? {
        ...metricAggregation,
        terms_agg: {
          terms: {
            field: groupBy[0],
          },
        },
      }
    : metricAggregation;

  return {
    over: {
      date_histogram: {
        field: timeField,
        interval: `${bucketValue}${bucketUnitOfTime}`,
        time_zone: moment.tz.guess(),
        min_doc_count: 0,
        extended_bounds: {
          min: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          max: 'now',
        },
      },
      aggregations,
    },
  };
}

export function formikToWhereClause({ filters = [] }) {
  return filters.map((filter) => OPERATORS_QUERY_MAP[filter.operator].query(filter));
}

// For query-level monitor single metric selection
export function formikToMetricAggregation(values) {
  const { aggregations } = values;
  let aggregationType = aggregations[0]?.aggregationType;
  const field = aggregations[0]?.fieldName;
  if (!field) return {};
  if (aggregationType === 'count') aggregationType = 'value_count';
  return { metric: { [aggregationType]: { field } } };
}

export function formikToUiCompositeAggregation(values) {
  const {
    aggregations,
    groupBy,
    timeField,
    bucketValue,
    bucketUnitOfTime,
    monitor_type: monitorType,
  } = values;

  let aggs = {};
  aggregations.map((aggItem) => {
    // TODO: Changing any occurrence of '.' in the fieldName to '_' since the
    //  bucketSelector uses the '.' syntax to resolve aggregation paths.
    //  Should revisit this as replacing with `_` could cause collisions with fields named like that.
    const name = `${aggItem.aggregationType}_${aggItem.fieldName.replace(/\./g, '_')}`;
    const type = aggItem.aggregationType === 'count' ? 'value_count' : aggItem.aggregationType;
    aggs[name] = {
      [type]: { field: aggItem.fieldName },
    };
  });

  let sources = [];
  groupBy.map((groupByItem) =>
    sources.push({
      [groupByItem]: {
        terms: {
          field: groupByItem,
        },
      },
    })
  );
  sources.push({
    date: {
      date_histogram: {
        field: timeField,
        interval: `${bucketValue}${bucketUnitOfTime}`,
        time_zone: moment.tz.guess(),
      },
    },
  });
  return {
    composite_agg: {
      composite: {
        sources,
        size: DEFAULT_COMPOSITE_AGG_SIZE,
      },
      aggs,
    },
  };
}

export function formikToUiSchedule(values) {
  return {
    timezone: _.get(values, 'timezone[0].label', null),
    frequency: values.frequency,
    period: values.period,
    daily: values.daily,
    weekly: values.weekly,
    monthly: values.monthly,
    cronExpression: values.cronExpression,
  };
}

export function buildSchedule(scheduleType, values) {
  const {
    period,
    daily,
    weekly,
    monthly: { type, day },
    cronExpression,
    timezone,
  } = values;
  switch (scheduleType) {
    case 'interval': {
      return { period };
    }
    case 'daily': {
      return { cron: { expression: `0 ${daily} * * *`, timezone } };
    }
    case 'weekly': {
      const daysOfWeek = Object.entries(weekly)
        .filter(([day, checked]) => checked)
        .map(([day]) => day.toUpperCase())
        .join(',');
      return { cron: { expression: `0 ${daily} * * ${daysOfWeek}`, timezone } };
    }
    case 'monthly': {
      let dayOfMonth = '?';
      if (type === 'day') {
        dayOfMonth = day;
      }
      return { cron: { expression: `0 ${daily} ${dayOfMonth} */1 *`, timezone } };
    }
    case 'cronExpression':
      return { cron: { expression: cronExpression, timezone } };
  }
}
