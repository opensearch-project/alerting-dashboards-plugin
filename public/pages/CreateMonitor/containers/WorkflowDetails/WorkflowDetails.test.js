/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from 'enzyme';

import { Formik } from 'formik';
import { FORMIK_INITIAL_VALUES } from '../CreateMonitor/utils/constants';
import WorkflowDetails from './WorkflowDetails';

describe('WorkflowDetails', () => {
  test('renders', () => {
    const component = (
      <Formik initialValues={FORMIK_INITIAL_VALUES} onSubmit={() => {}}>
        <WorkflowDetails isDarkMode={false} values={{}} httpClient={{}} errors={{}} />
      </Formik>
    );
    expect(render(component)).toMatchSnapshot();
  });
});
