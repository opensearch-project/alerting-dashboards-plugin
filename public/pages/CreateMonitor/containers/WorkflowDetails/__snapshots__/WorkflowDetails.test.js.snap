// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowDetails renders 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
  style="padding-left:10px;padding-right:10px;padding-bottom:20px;padding-top:20px"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding:0px 10px"
  >
    <div
      class="euiFlexItem"
    >
      <div
        class="euiText euiText--small"
      >
        <h2>
          Workflow
        </h2>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        />
      </div>
    </div>
  </div>
  <div
    class="euiText euiText--extraSmall"
    style="padding:0px 10px"
  >
    <div
      class="euiTextColor euiTextColor--subdued"
    />
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding:0px 10px"
  >
    <div
      class="euiText euiText--medium"
      style="margin-bottom:0px"
    >
      <h4>
        Schedule
      </h4>
    </div>
    <div
      class="euiSpacer euiSpacer--m"
    />
    <div
      style="max-width:400px"
    >
      <div>
        <div
          class="euiFormRow"
          id="frequency-form-row-row"
        >
          <div
            class="euiFormRow__labelWrapper"
          >
            <label
              aria-invalid="false"
              class="euiFormLabel euiFormRow__label"
              for="frequency-form-row"
            >
              Frequency
            </label>
          </div>
          <div
            class="euiFormRow__fieldWrapper"
          >
            <div
              class="euiFormControlLayout"
            >
              <div
                class="euiFormControlLayout__childrenWrapper"
              >
                <select
                  class="euiSelect"
                  data-test-subj="frequency_field"
                  id="frequency"
                  name="frequency"
                >
                  <option
                    selected=""
                    value="interval"
                  >
                    By interval
                  </option>
                  <option
                    value="daily"
                  >
                    Daily
                  </option>
                  <option
                    value="weekly"
                  >
                    Weekly
                  </option>
                  <option
                    value="monthly"
                  >
                    Monthly
                  </option>
                  <option
                    value="cronExpression"
                  >
                    Custom cron expression
                  </option>
                </select>
                <div
                  class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                >
                  <span
                    class="euiFormControlLayoutCustomIcon"
                  >
                    <div>
                      EuiIconMock
                    </div>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiSpacer euiSpacer--s"
        />
        <div
          class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow"
              id="period.interval-form-row-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  aria-invalid="false"
                  class="euiFormLabel euiFormRow__label"
                  for="period.interval-form-row"
                >
                  Run every
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <input
                      class="euiFieldNumber euiFieldNumber--withIcon"
                      data-test-subj="interval_interval_field"
                      min="1"
                      name="period.interval"
                      type="number"
                      value="1"
                    />
                    <div
                      class="euiFormControlLayoutIcons"
                    >
                      <span
                        class="euiFormControlLayoutCustomIcon"
                      >
                        <div>
                          EuiIconMock
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFormRow euiFormRow--hasEmptyLabelSpace"
              id="period.unit-form-row-row"
            >
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class="euiFormControlLayout"
                >
                  <div
                    class="euiFormControlLayout__childrenWrapper"
                  >
                    <select
                      class="euiSelect"
                      data-test-subj="interval_unit_field"
                      id="period.unit"
                      name="period.unit"
                    >
                      <option
                        selected=""
                        value="MINUTES"
                      >
                        Minute(s)
                      </option>
                      <option
                        value="HOURS"
                      >
                        Hours
                      </option>
                      <option
                        value="DAYS"
                      >
                        Days
                      </option>
                    </select>
                    <div
                      class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                    >
                      <span
                        class="euiFormControlLayoutCustomIcon"
                      >
                        <div>
                          EuiIconMock
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--xl"
    />
    <h4
      class="euiTitle euiTitle--xsmall"
    >
      Delegate monitors
    </h4>
    <div
      class="euiText euiText--extraSmall"
    >
      <div
        class="euiTextColor euiTextColor--subdued"
      >
        Delegate two or more monitors to run as part of this workflow. The order in which you select the monitors determines their order in the workflow. The monitor types per query, per bucket, and per document are supported.Â 
        <a
          class="euiLink euiLink--primary"
          href="https://opensearch.org/docs/latest/observing-your-data/alerting/monitors/#monitor-types"
          rel="noopener noreferrer"
          target="_blank"
        >
          Learn more.
          <div>
            EuiIconMock
          </div>
          <span
            class="euiScreenReaderOnly"
          >
            (opens in a new tab or window)
          </span>
        </a>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--m"
    />
    <div
      class="euiFormRow euiFormRow--fullWidth"
      id="associatedMonitorsEditor-form-row-row"
    >
      <div
        class="euiFormRow__labelWrapper"
      >
        <label
          class="euiFormLabel euiFormRow__label"
          for="associatedMonitorsEditor-form-row"
        >
          Define workflow
        </label>
      </div>
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiCodeEditorWrapper"
          data-test-subj="associatedMonitorsEditor"
          style="width:80%;height:300px"
        >
          <button
            class="euiCodeEditorKeyboardHint"
            data-test-subj="codeEditorHint"
            id="generated-id"
          >
            <p
              class="euiText"
            >
              Press Enter to start editing.
            </p>
            <p
              class="euiText"
            >
              When you're done, press Escape to stop editing.
            </p>
          </button>
          <div
            id="generated-id"
            style="width:80%;height:300px"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
